src/java.base/share/classes/jdk/crac/Core.java

```diff
@@ -247,23 +251,34 @@ public static void checkpointRestore() throws
         }
     }
 
-    /* called by VM */
-    private static void checkpointRestoreInternal() {
-        Thread thread = new Thread(() -> {
-            try {
-                Thread.sleep(100);
-            } catch (InterruptedException e) {
+    private static void checkpointRestore2(long outputStream_p, long jcmd_p) throws
+            CheckpointException,
+            RestoreException {
+        synchronized (checkpointRestoreLock) {
+            // idkn, it worth to protect from multy-thread checkpointing ?  - doesn't have sence for jcmd checkpointing, only for
+            // jcmd + somebody through an api ...
+            if (!checkpointInProgress) {
+                checkpointInProgress = true;
+                try {
+                    checkpointRestore1(outputStream_p, jcmd_p);
+                } finally {
+                    checkpointInProgress = false;
+                }
             }
+        }
```

Comment:
Let's move the whole recursive check from public checkpointRestore() here, and make that only to call the private method.